### DISCUSSION TITLE
Cloning the wrong layer?

### DISCUSSION THREAD

**Discussion by aarya123** [ImageMagick/ImageMagick#3738] [Help] [answered]
So, I'm trying to generate an image with a rounded corner, but, for some reason, the overlay of the logo results in the wrong layer being cloned. I've tried adding `-layers flatten` and `-layers merge` after the overlay is composited, but it seems to not be cloning the right layer. I've also tried `-delete 1` after every composite to keep the stack as flat as possible, but no luck. Command:

```
magick source.jpeg -resize 548x308! \
    \( -size 548x308 -define gradient:vector="126,308 367,0" gradient:"rgba(51,136,153,1.000)-rgba(51,136,153,0.000)" \) -composite \
    \( -size 548x308 -define gradient:vector="274,308 274,71" gradient:"rgba(0,0,0,0.700)-rgba(0,0,0,0.000)" \) -composite \
    \( -size 548x308 -define gradient:vector="0,308 279,148" gradient:"rgba(51,136,153,0.300)-rgba(51,136,153,0.000)" \) -composite \
    \( overlay.png -resize 80x80 -gravity Center -background "rgba(0,0,0,0.000)" -extent 80x80 \) -geometry +26+26 -gravity SouthEast -composite \
    \( +clone  -alpha extract -draw 'fill black polygon 0,0 0,9 9,0 fill white circle 9,9 9,0' \
        \( +clone -flip \) -compose Multiply -composite \
        \( +clone -flop \) -compose Multiply -composite \
     \) -alpha off -compose CopyOpacity -composite -alpha associate \
      -colorspace sRGB -format png -quality 75.0 -strip out.png
```
[source.jpeg](https://user-images.githubusercontent.com/882531/119414225-0507fc80-bca4-11eb-8a41-e7dc8ddefdca.jpeg)
[overlay.png](https://user-images.githubusercontent.com/882531/119414224-046f6600-bca4-11eb-859a-5461a6b962e0.png)
What I get: [output.png](https://user-images.githubusercontent.com/882531/119414220-033e3900-bca4-11eb-8392-efef8fe7d1bb.png)
What I want I want with the overlay: [RoundedCorners](https://user-images.githubusercontent.com/882531/119414500-852e6200-bca4-11eb-9f6f-d678d49e5fb9.png)

```
➜  test magick -version                        
Version: ImageMagick 7.0.11-9 Q16 x86_64 2021-04-24 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib
```

**Comment by fmw42** [comment]
Why are you using overlay.png to get rounded corners? For rounded corners see for example https://legacy.imagemagick.org/Usage/thumbnails/#rounded

```
convert 119414225-0507fc80-bca4-11eb-8a41-e7dc8ddefdca.jpeg \
\( +clone  -alpha extract \
-draw 'fill black polygon 0,0 0,150 150,0 fill white circle 150,150 150,0' \
\( +clone -flip \) -compose Multiply -composite \
\( +clone -flop \) -compose Multiply -composite \
\) -alpha off -compose CopyOpacity -composite  rounded_corners.png

```

![rounded_corners](https://user-images.githubusercontent.com/13458218/119416939-9a59bf80-bca9-11eb-8d62-1ec4c5c65fe7.png)


In ImageMagick 7, you can get relative size rounding as follows:

 ```
magick 119414225-0507fc80-bca4-11eb-8a41-e7dc8ddefdca.jpeg \
\( +clone  -alpha extract \
-draw 'fill black polygon 0,0 0,%[fx:0.05*max(w,h)] %[fx:0.05*max(w,h)],0 fill white circle %[fx:0.05*max(w,h)],%[fx:0.05*max(w,h)] %[fx:0.05*max(w,h)],0' \
\( +clone -flip \) -compose Multiply -composite \
\( +clone -flop \) -compose Multiply -composite \
\) -alpha off -compose CopyOpacity -composite  rounded_corners.png
```

**Reply by aarya123** [reply]
No, overlay.png is something I'd like to add to my image prior to rounding the corners, along with a few other things, like gradients. However, it seems like I can't get `+clone` to clone the entire composited image.

And yea, those are the docs I'm modeling off of. I have my command working with everything, except for when I add in the overlay. So even with just the gradients, it looks fine

**Reply by fmw42** [reply]
I am not sure I understand. Does it work if you replace +clone with -clone 0? Which +clone is not working?

**Reply by aarya123** [reply]
Oh, I assume it's the first `+clone` because that's working off of the composited set of images. The subsequent ones are working off of the new image I'm drawing given that it's a nested operation.
```
magick source.jpeg -resize 548x308! \
    \( -size 548x308 -define gradient:vector="126,308 367,0" gradient:"rgba(51,136,153,1.000)-rgba(51,136,153,0.000)" \) -composite \
    \( -size 548x308 -define gradient:vector="274,308 274,71" gradient:"rgba(0,0,0,0.700)-rgba(0,0,0,0.000)" \) -composite \
    \( -size 548x308 -define gradient:vector="0,308 279,148" gradient:"rgba(51,136,153,0.300)-rgba(51,136,153,0.000)" \) -composite \
    \( overlay.png -resize 80x80 -gravity Center -background "rgba(0,0,0,0.000)" -extent 80x80 \) -geometry +26+26 -gravity SouthEast -composite \
    \( -clone 0 -alpha extract -draw 'fill black polygon 0,0 0,9 9,0 fill white circle 9,9 9,0' \
        \( +clone -flip \) -compose Multiply -composite \
        \( +clone -flop \) -compose Multiply -composite \
     \) -alpha off -compose CopyOpacity -composite -alpha associate \
      -colorspace sRGB -format png -quality 75.0 -strip out.png
```
^This did not work either

**Reply by fmw42** [reply]
Your +clone -alpha extract is cloning the result of the composite not the overlay extended to 80x80. What is you want for that clone? Once you do -composite, you cannot get back to the images used by the composite unless you save them as clones or as mpr: format.

**Reply by aarya123** [reply]
Oh, interesting, so there goes the cloning being the issue theory. So I guess, given this image [here](https://user-images.githubusercontent.com/882531/119419084-4e5d4980-bcae-11eb-8f11-24782d9118b1.png), generated by:
```
magick source.jpeg -resize 548x308! \
    \( -size 548x308 -define gradient:vector="126,308 367,0" gradient:"rgba(51,136,153,1.000)-rgba(51,136,153,0.000)" \) -composite \
    \( -size 548x308 -define gradient:vector="274,308 274,71" gradient:"rgba(0,0,0,0.700)-rgba(0,0,0,0.000)" \) -composite \
    \( -size 548x308 -define gradient:vector="0,308 279,148" gradient:"rgba(51,136,153,0.300)-rgba(51,136,153,0.000)" \) -composite \
    \( overlay.png -resize 80x80 -gravity Center -background "rgba(0,0,0,0.000)" -extent 80x80 \) -geometry +26+26 -gravity SouthEast -composite \
      -colorspace sRGB -format png -quality 75.0 -strip out.png
```
How can I round the corners via a suboperation? The mask that I'm generating doesn't seem to work:
```
    \( +clone -alpha extract -draw 'fill black polygon 0,0 0,9 9,0 fill white circle 9,9 9,0' \
        \( +clone -flip \) -compose Multiply -composite \
        \( +clone -flop \) -compose Multiply -composite \
     \) -alpha off -compose CopyOpacity -composite -alpha associate \
```
This should be the end result: 
![out](https://user-images.githubusercontent.com/882531/119419274-c9266480-bcae-11eb-9178-a8a30bdcdb46.png)

**Comment by fmw42** [comment, marked as answer]
Your corners are being drawn but just too small. Also you need to turn off -geometry after using it by adding +geometry, so that the offset is not being used by subsequent commands. So this works after removing the last bit. -quality 75 is not correct for PNG. I do not know why you are adding -alpha associate nor -strip nor -colorspace sRGB. Your input is already sRGB.

```
magick 119414225-0507fc80-bca4-11eb-8a41-e7dc8ddefdca.jpeg -resize 548x308! \
\( -size 548x308 -define gradient:vector="126,308 367,0" gradient:"rgba(51,136,153,1.000)-rgba(51,136,153,0.000)" \) -composite \
\( -size 548x308 -define gradient:vector="274,308 274,71" gradient:"rgba(0,0,0,0.700)-rgba(0,0,0,0.000)" \) -composite \
\( -size 548x308 -define gradient:vector="0,308 279,148" gradient:"rgba(51,136,153,0.300)-rgba(51,136,153,0.000)" \) -composite  \
\( 119414224-046f6600-bca4-11eb-859a-5461a6b962e0.png -resize 80x80 -gravity Center -background "rgba(0,0,0,0.000)" -extent 80x80  \) -geometry +26+26 -gravity SouthEast -composite +geometry \
\( +clone -alpha extract -draw 'fill black polygon 0,0 0,90 90,0 fill white circle 90,90 90,0'  \
\( +clone -flip \) -compose Multiply -composite \
\( +clone -flop \) -compose Multiply -composite \
\) -alpha off -compose CopyOpacity -composite \
out.png
```

![out](https://user-images.githubusercontent.com/13458218/119420028-58804780-bcb0-11eb-8a09-49421c20d99c.png)

Change the rounding amount as desired or use %[fx:max(w,h)*factor]

**Reply by aarya123** [reply]
Ah, `+geometry` did the trick! And yea the colorspace, strip, and quality are unnecessary. Just forgot to remove them since I was working with a separate set of permutations for jpeg.
The reason why I'm adding -alpha associate is so that the transparent pixels are cleared out. For example, if I do the following:
```
magick source.jpeg -resize 1920x1080! \
    \( +clone  -alpha extract -draw 'fill black polygon 0,0 0,960 960,0 fill white circle 960,540 960,0' \
        \( +clone -flip \) -compose Multiply -composite \
        \( +clone -flop \) -compose Multiply -composite \
    \) -alpha off -compose CopyOpacity -composite -alpha associate \
    -format png rounded6.png
```
and
```
magick source.jpeg -resize 1920x1080! \
    \( +clone  -alpha extract -draw 'fill black polygon 0,0 0,960 960,0 fill white circle 960,540 960,0' \
        \( +clone -flip \) -compose Multiply -composite \
        \( +clone -flop \) -compose Multiply -composite \
    \) -alpha off -compose CopyOpacity -composite \
    -format png rounded6.png
```
Difference being `-alpha associate`, you'll notice that the file sizes changes drastically.

**Reply by aarya123** [reply]
Thanks for the help, @fmw42! :)
